### PR TITLE
ENH: changes for convolution symmetry

### LINKED ISSUE #5897: Order of convolve inputs shouldn't matter
Reading through https://github.com/scipy/scipy/pull/5608 I noticed that the docstring still says "if sizes of `in1` and `in2` are not equal then `in1` has to be the larger array", but [convolution is commutative](https://en.wikipedia.org/wiki/Convolution#Properties), so it shouldn't matter at all which order the inputs are in, yet it does because it goes through `_check_valid_mode_shapes()`:

```
convolve([2, 3, 4, 5], [1], 'full')
Out[45]: array([2, 3, 4, 5])

convolve([1], [2, 3, 4, 5], 'full')
Out[46]: array([2, 3, 4, 5])

convolve([2, 3, 4, 5], [1], 'valid')
Out[47]: array([2, 3, 4, 5])

convolve([1], [2, 3, 4, 5], 'valid')
Traceback (most recent call last):
...
ValueError: in1 should have at least as many items as in2 in every dimension for 'valid' mode.
```

I changed this partially in https://github.com/scipy/scipy/pull/3664 but I guess the `swapped_inputs` logic should apply to the entire function, and drop the `valid` mode size check?

I don't see why this should restrict `correlate`, either.  I think `valid` mode still makes sense regardless of which input is larger?  The output is just reversed and conjugated if you swap the inputs?

### PR DESCRIPTION
**Pull request by endolith**
I started to fix https://github.com/scipy/scipy/issues/5897 before I realized it had already been done in https://github.com/scipy/scipy/pull/5904, so these are the other changes I made in the process.

(I was going to change `slice_obj = [slice(None, None, -1)] * len(kernel.shape)` into `kernel[::-1]` before I remembered it's N-dimensional, so I added a comment and renamed the variable to clarify that for others.)

`real_array.conj()` does nothing and just returns `real_array`, faster than
`iscomplexobj(real_array)` tests whether it's real, so I changed it to just call `.conj()` in both cases.

```
a = rand(4)

a.conj() is a
Out[105]: True
```

also changed a bunch of `len(var.shape)` into `var.ndim`.

### CODE REVIEW

**Comment by gfyoung** [comment]
+1 for the changes.  This is probably minor, but merging mine before yours seems slightly preferable since rebasing your changes onto my PR would be a little easier than the other way IMHO.

File: scipy/signal/signaltools.py [hunk 1]
```diff
@@ -470,12 +471,10 @@ def convolve(in1, in2, mode='full'):
...
 
-    slice_obj = [slice(None, None, -1)] * len(kernel.shape)
+    # Reverse in all dimensions
+    reverse = [slice(None, None, -1)] * kernel.ndim
 
-    if np.iscomplexobj(kernel):
-        return correlate(volume, kernel[slice_obj].conj(), mode)
-    else:
-        return correlate(volume, kernel[slice_obj], mode)
+    return correlate(volume, kernel[reverse].conj(), mode)
```
Comment (Eric89GXL) [inline scipy/signal/signaltools.py]: Can you add a comment that using `.conj()` is faster than checking `iscomplexobj` even if it's not necessary to do it (for real inputs), like you mentioned in the comments of this PR?

**Comment by Eric89GXL** [comment]
LGTM, +1 for merge other than my comment. @endolith you okay with taking one for the team and waiting for @gfyoung's PR to land so you can rebase?

**Comment by endolith** [comment]
added the comment.

I don't care when it gets merged :)

also I could rebase it on top of @gfyoung 's PR and make it part of that PR

**Comment by gfyoung** [comment]
Rebasing immediately is probably fine, though I'm hoping my PR can get merged in soon and then it will be easier (at least in my mind) for you to rebase onto `master` instead. :smile:

**Comment by endolith** [comment]
added some more things to make sure they aren't forgotten about

**Comment by endolith** [comment]
Merged into https://github.com/scipy/scipy/pull/5904

**Comment by endolith** [comment]
Looks like these didn't make it into https://github.com/scipy/scipy/pull/5904 so I'll reopen this PR

**Comment by endolith** [comment]
(also making some other changes...)

**Comment by endolith** [comment]
I don't know when this was introduced, but there's a discrepancy in handling of complex conjugates in scipy 0.17.0:

```
    scipy.signal.correlate(1j, 2j)              (-2+0j)
    scipy.signal.correlate([1j], [2j])          array([ 2.+0.j])
    scipy.signal.correlate([[1j]], [[2j]])      array([[ 2.+0.j]])
    scipy.signal.correlate2d([[1j]], [[2j]])    array([[-2.+0.j]]) 
    numpy.correlate([1j], [2j])                 array([ 2.+0.j])
```

Wikipedia's definition of cross-correlation is 

df665b17d676571c9dc7a1800e1b186a 1

which would mean conjugating the 1j into -1j and then multiplying it to get +2 as the correct output.

**Comment by gfyoung** [comment]
@endolith : Not really sure why your changes didn't get merged in when my PR was merged.  Sorry about that!  Your efforts did not go unappreciated though!

**Comment by e-q** [comment]
LGTM, thanks @endolith !

**Pull request merged**